### ISSUE TITLE
Consider updating mutability syntax for storage accesses

### ISSUE THREAD

**Issue by mohammadfawaz** [essential-contributions/pint#1021]
Current syntax is a little clunky IMO. It's fine and does the job, but it feels to me that the `mut` keywords are "burried" in the program. A solver would have to go hunt them down to figure out what keys are mutable.

We discussed a few alternatives. One important requirement is that the mutable storage expressions must have access to the body of the predicate and its arguments so they must live inside the body of the predicate. Here's one suggestion:

```rust
storage {
    balance: (b256 => int),
    value: int,
}

predicate First(key2: b256, key2: b256, pub_key: { b256, b256 }) {
    let key3 = __sha256(pub_key);
    mutate storage::{
        balance[key1], 
        balance[key2],
        balance[key3],
        value,
    };

    let sum = storage::balance[key1]' + storage::balance[key2]'  + storage::balance[key3]' + value;
    ...
}
```

Here, a `mutate` declaration is a bit like a `use` declaration (similar syntax). It tells us that the predicate mutates these keys. There are two downsides:
1. Repetition of the storage expressions.
2. Can have multiple `mutate` declarations but we can require that only one exists.
3. The `storage` keyword used in the `mutate` declaration. Technically, it's not required though because nothing else can go there. We can make it optional so that people can add it for clarity but can be skipped for conciseness.

**Comment by nfurfaro** [comment]
Another option: we come up with a syntax to say that a storage slot is mutable by a given predicate/predicates.
Something like:

```
storage {
    mut balances: (b256 => int) for First, // or balances: (b256 => int) mut<First>, 
    value: int mut(Second),
}

predicate First(key1: b256, key2: b256, pub_key: { b256, b256 }) {
    let key3 = __sha256(pub_key);
    let sum = storage::balances[key1]' + storage::balances[key2]'  + storage::balances[key3]' + value;    
}

predicate Second() {
    let val: int = storage::value;
    constraint val < val';
}
```
Here, we don't specifically say that `storage::balances[key1]` is mutable, but rather that `First` may mutate `storage::balances` , and the body of `First` specifies which specific keys it may mutate.

I'm not sure if there are issues here with the mutablility "leaking" and introducing unintended mutablility with this approach... that is, I don't know if we can declare a storage slot mutable only by certain predicates?

**Comment by otrho** [comment]
> Another option: we come up with a syntax to say that a storage slot is mutable by a given predicate/predicates. Something like:
> 
> ```
> storage {
>     mut balances: (b256 => int) for First, // or balances: (b256 => int) mut<First>, 
>     value: int mut(Second),
> }
> ```

Yeah, I don't think this is fine grained enough.  We might only want a single key in the map to be mutable, or for a single element in an array.

**Comment by otrho** [comment]
> We discussed a few alternatives. One important requirement is that the mutable storage expressions must have access to the body of the predicate and its arguments so they must live inside the body of the predicate. Here's one suggestion:
>

This is probably the best so far, but I'm not a fan of it not being a part of the signature.  We don't have shadowing so there is no order to declarations, so in theory there's nothing wrong with:

```
storage {
    balances: (b256 => int),
    value: int,
}

predicate First(key1: b256, key2: b256, pub_key: { b256, b256 })
  mutates (
    balances[key1],
    balances[key2],
    balances[key3],
    value,
) {
    let key3 = __sha256(pub_key);

    ...
}
```

All hail `mutates`!  Although `key3` is in the body it's unambiguous.  It's just declared in the body and anyone reading it would have to assume as much.  IMO it's not that different to out of order `let` decls.

**Comment by mohammadfawaz** [comment]
yeah I don't mind this either
